Add Alias implementation to the DG (#280)
AY2122S1-CS2103T-W08-2/tp: docs/DeveloperGuide.md

    * Pros: More straightforward and convenient for users.
    * Cons: We need to identify the type of input given.

### Alias feature

The `alias` feature allows users to define their own aliases for commands. This is useful to shorten the input for commands that the user uses often. 

> An alias is basically just a mapping of a user-provided string to a command word. It works by directly replacing the first word in the argument with the alias word, then parsing again.

Reviewer: ```suggestion
An alias is basically just a mapping of a user-provided string to a command word. It works by directly replacing the first word in the user's input command with the alias word, then parsing again.
```

> An alias contains two strings: An `aliasWord` which is the new user-defined word, and an `commandWord` which is the command word of an existing default command (e.g. `addstudent`, `clear`, `exit`).

Reviewer: ```suggestion
An alias contains two strings: An `aliasWord` which is the new user-defined word, and a `commandWord` which is the command word of an existing default command (e.g. `addstudent`, `clear`, `exit`).
```

Parsing of an alias command follows the following steps:
1. The alias word is checked to ensure that it is one word long.

> 1. The alias word is also check to ensure that it does not overlap with any default command. This is to prevent the remapping of any default commands and potentially losing the functionality of the application.

Reviewer: ```suggestion
1. The alias word is also checked to ensure that it does not overlap with any default command word. This is to prevent the re-mapping of any default command words and potentially losing the functionality of the application.
```

1. If the command word is an alias, it is replaced with the command word that the alias maps to.
1. The command word is checked for validity by attempting to parse the command word. If the parser does not recognise the command word, it is not valid.
1. The new `Alias` and `AliasCommand` is created, and executed.
1. The alias is added to both the parser, and the model (so that it can be saved in the `UserPrefs`)
    * If the alias word is already present in the parser, the command word it is mapped to is replaced with the new command word instead.

>     * If the alias word and the command word are identical, the existing alias is removed instead.

Reviewer: ~I'm not sure what this means, the way it's phrased makes it sound like the second step? (I know if it was the second step it would have terminated then, but this isn't really clear)~

After reading to the end, I now get what it means. 
```suggestion
    * If the alias word and the command word are identical, it is assumed that the user is trying to delete the existing alias. The existing alias is removed.
```

    
If the command word of any future user input matches the alias, the first word of the user input will be replaced by the command word of the alias.
The activity diagram of how a command is parsed can be found in the [Logic section of this guide](#logic-component).

#### Design Considerations

> **Aspect 1: How aliases are stored and how they parse**

Reviewer: ```suggestion
**Aspect 1: How aliases are stored and parsed**
```

There were two ideas on how this could be done:
1. Store the command's `Parser` in each alias. The parser attribute of the alias can then be retrieved and `Parser::execute` can be called when parsing.
    * Pros:

>         * Very elegant and streamlined parsing. Almost no changes need to be made to the logic of `AddressBookParser`.

Reviewer: ```suggestion
        * Very elegant and streamlined parsing. Almost no changes need to be made to the logic of `SourceControlParser`.
```

    * Cons:
        * It is not easy to obtain the type of `Parser` simply through the command word provided by the user given the current implementation of the project. You'd need to iterate through all possible `Commands`, and match their command word with the provided word. This approach would not be very extensible if more commands are to be added in the future. 

>         * It is also not easy to save/load the alias to/from the `UserPrefs`. The best approach is likely to just save the CommandWord, then generate the Parser when loading. However this poses a similar problem as the previous point, which is that it is not easy to generate the `Parser` just from the command word.

Reviewer: ```suggestion
        * It is also not easy to save/load the alias to/from the `UserPrefs`. The best approach is likely to just save the command word, then generate the `Parser` when loading. However this poses a similar problem as the previous point, which is that it is not easy to generate the `Parser` just from the command word.
```

1. Store the command word in each alias. Replace the command word of the user input if the user inputs an aliased command, and parse it again.
    * Pros:
        * Very easy to store. Each alias is essentially just two strings.
    * Cons: 
        * Have to be careful with the implementation to prevent any edge cases where the user can define aliases to create an infinite loop.
    
We decided to go with the easier implementation of storing each alias as two strings. However, there were still more aspects to be considered.

**Aspect 2: How to handle aliases of aliases**

That is, what happens when the user does `alias -c <existing_alias> -as <new_alias>`? There are two choices for this: 
1. The new alias maps to the existing alias word directly.
    * The command word stored in the new `Alias` would be the alias word of the existing `Alias`.
    * Assuming that an alias `as` already exists for `addstudent`, the following commands have the following effects:

>         * `alias -c as -as as2` would make `as2` map to `addstudent` via `as`.

Reviewer: ```suggestion
        * `alias -c as -as as2` would make `as2` map to `as`, which will be then mapped to `addstudent` via `as` when used.
```
Would this better convey the meaning? I think the "make `as2` map to `addstudent`" may trip up the reader

>         * `alias -c addscore -as as` would make `as` map to `addscore`. `as2` would also then map to `addscore`, since it gets replaced by `as` when parsing, which then gets replaced by `addscore`.

Reviewer: Would it be better if we use another example like `ag` -> `addgroup`? I'm afraid the user may be confused like in https://github.com/AY2122S1-CS2103T-W08-2/tp/issues/236

    * Pros:
        * Intuitive approach to aliases: "I want `as2` to mean `as`".
    * Cons:
        * Possible to make parsing a very expensive task due to having to replace the command word and parsing multiple times.
        * Possible to have an infinite loop by mapping two aliases to each other.
        * It's not very useful to have two aliases always mean the same thing.

> 2. The new alias maps to the command that the existing alias maps to.

Reviewer: ```suggestion
2. The new alias maps to the command word that the existing alias maps to.
```
